src/networking.c

```diff
@@ -4316,44 +4317,47 @@ void killIOThreads(void) {
-    /* We may have still clients with pending reads when this function
-     * is called: handle them before stopping the threads. */
-    handleClientsWithPendingReadsUsingThreads();
-    serverAssert(server.io_threads_active == 1);
-    for (int j = 1; j < server.io_threads_num; j++)
-        pthread_mutex_lock(&io_threads_mutex[j]);
-    server.io_threads_active = 0;
-}
-
-/* This function checks if there are not enough pending clients to justify
- * taking the I/O threads active: in that case I/O threads are stopped if
- * currently active. We track the pending writes as a measure of clients
- * we need to handle in parallel, however the I/O threading is disabled
- * globally for reads as well if we have too little pending clients.
+/* This function dynamically adjusts the I/O thread number according to
+ * real workloads in runtime, currently we track the latest x rounds of
+ * pending writes as a measure of clients we need to handle in parallel,
+ * however the I/O threading is disabled globally for reads as well if
+ * we have too little pending clients.
  *
  * The function returns 0 if the I/O threading should be used because there
  * are enough active threads, otherwise 1 is returned and the I/O threads
  * could be possibly stopped (if already active) as a side effect. */
+#define IO_THREAD_BATCH_NUM 2
+#define SMOOTH_FACTOR 0.8
 int stopThreadedIOIfNeeded(void) {
-    int pending = listLength(server.clients_pending_write);
-
-    /* Return ASAP if IO threads are disabled (single threaded mode). */
+    /* Return ASAP if I/O threads are disabled (single threaded mode). */
     if (server.io_threads_num == 1) return 1;
 
-    if (pending < (server.io_threads_num*2)) {
-        if (server.io_threads_active) stopThreadedIO();
-        return 1;
+    server.clients_pending_write_avg_num = (int)(
+        server.clients_pending_write_avg_num * SMOOTH_FACTOR +
+        listLength(server.clients_pending_write) * (1 - SMOOTH_FACTOR));
+
+    /* Target io-thread number should be in range of [1, io_threads_num]. */
+    int targetNum = min(server.io_threads_num,
+        max(1, server.clients_pending_write_avg_num / IO_THREAD_BATCH_NUM));
+
+    if (server.io_threads_active_num < targetNum) {
+        /* Increasing active threads. */
+        for (int i = server.io_threads_active_num; i < targetNum; ++i) {
+            pthread_mutex_unlock(&io_threads_mutex[i]);
+            server.io_threads_active_num++;
+        }
     } else {
-        return 0;
+        /* We may have still clients with pending reads when this function
+         * is called: handle them before stopping the threads. */
+        handleClientsWithPendingReadsUsingThreads();
+        /* Decreasing active threads. */
+        for (int i = server.io_threads_active_num; i > targetNum; --i) {
+            pthread_mutex_lock(&io_threads_mutex[i - 1]);
+            server.io_threads_active_num--;
+        }
     }
+
+    return server.io_threads_active_num == 1;
```

Comment:
There are two options here: either return `server.io_threads_active_num` or return `void`. My preference would be returning the active IO thread count and this will remove all the references to the global var `server.io_threads_active_num` later on. I am always for less coupling, at any level.

```suggestion
    return server.io_threads_active_num;
```